pkg/operator/ceph/object/health.go

```diff
@@ -0,0 +1,272 @@
+			logger.Infof("stopping monitoring of rgw endpoints for object store %q", c.namespacedName.Name)
+			return
+
+		case <-time.After(defaultStatusCheckInterval):
+			logger.Debug("checking rgw health of object store %q", c.namespacedName.Name)
+			err := c.checkObjectStoreHealth()
+			if err != nil {
+				logger.Warningf("failed to check rgw health for object store %q. %v", c.namespacedName.Name, err)
+			}
+		}
+	}
+}
+
+func (c *bucketChecker) checkObjectStoreHealth() error {
+	/*
+		0. purge everything by default
+		1. create an S3 user
+		2. always use the same user
+		3. if already exists just re-hydrate the s3 credentials
+		4. create a bucket with that user or use the existing one (always use the same bucket)
+		5. create a check file
+		6. get the MD5 of the file
+		7. PUT the file
+		8. GET the file
+		9. compare MD5
+		10. delete object on bucket
+		11. update CR health status check
+
+		Always keep the bucket and the user for the health check, just do PUT and GET because bucket creation is expensive
+	*/
+
+	var s3AccessKey string
+	var s3SecretKey string
+	s3endpoint := fmt.Sprintf("%s:%s", BuildDomainName(c.objContext.Name, c.namespacedName.Namespace), c.port)
+
+	// Generate unique user and bucket name
+	bucket := genUniqueBucketName(c.objContext.UID)
+	userConfig := c.genUserConfig()
+
+	// Create S3 user
+	logger.Debugf("creating s3 user object %q for object store %q", userConfig.UserID, c.namespacedName.Name)
+	user, rgwerr, err := CreateUser(c.objContext, userConfig)
+	if err != nil {
+		if rgwerr == ErrorCodeFileExists {
+			user, _, err = GetUser(c.objContext, userConfig.UserID)
+			if err != nil {
+				err = errors.Wrapf(err, "failed to get details from ceph object user %q for object store %q", user.UserID, c.namespacedName.Name)
+				updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+				return err
+			}
+		} else {
+			err = errors.Wrapf(err, "failed to create object user %q. error code %d for object store %q", userConfig.UserID, rgwerr, c.namespacedName.Name)
+			updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+			return err
+		}
+	}
+	// Set access and secret key
+	s3AccessKey = *user.AccessKey
+	s3SecretKey = *user.SecretKey
+
+	// Initiate s3 agent
+	logger.Debugf("initializing s3 connection for object store %q", c.namespacedName.Name)
+	s3client := CreateNewS3Client(s3endpoint, s3AccessKey, s3SecretKey)
+
+	// Force purge the s3 object before starting anything
+	err = cleanupObjectHealthCheck(s3client, c.objContext.UID)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to perform object cleanup for object store %q", c.namespacedName.Name)
+		updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+		return err
+	}
+
+	// Bucket health test
+	err = c.testBucketHealth(s3client, bucket)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to run bucket health checks for object store %q", c.namespacedName.Name)
```

Comment:
For example, just return the err and the caller will update the status on the CR